### PR TITLE
APEXCORE-222 Purging the stale data present in the bufferserver, dire…

### PR DESCRIPTION
**Pull request by sandeshh**
…ctly from the streaming container.

### CODE REVIEW

File: engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java [hunk 1]
```diff
@@ -769,7 +769,11 @@ public void processHeartbeatResponse(ContainerHeartbeatResponse rsp)
     }
 
     if (rsp.committedWindowId != lastCommittedWindowId) {
+
       lastCommittedWindowId = rsp.committedWindowId;
+
+      bufferServer.purge(lastCommittedWindowId);
```
Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: Shouldn't this be a request that is sent to bufferserver from the streaming container otherwise this introduces a dependency which we didn't rely on before that bufferserver is in the same process.

Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: I don't think we have any plans to run BufferServer in different process/container. There is Jira to replace the loopback  communication between BufferServer and the operators with a queue. So considering that points, I used this approach.

Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: I would think it would be an option in that case too. Do you know if there was a discussion to dev list to fundamentally change the approach and only do inline. If it was discussed and this conclusion was reached I think it would be prudent to wait for that change to get in before merging this one.

Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: That loopback communication is for the tuple flow. This change is not dependent on that. I can still make this a loopback from the streaming container.

Comment (pramodin) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: Today there is flexibility for the server to not be in the same process even though it is not being exercised. With this change that flexibility will be gone. How about providing both options.

Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: While currently buffer server by itself does not make any assumption regarding locality and communicates with its clients (publisher, subscriber, controller) over network request API, Apex supports deployment of the buffer server into the same container as the upstream operator/publisher only. As the buffer server is not fault tolerant, and there is no plan to make it fault tolerant, in reality, there is no flexibility to deploy buffer server into another container other than the upstream operator container. Additionally, as long as the buffer server continues to support purge request submitted over a network, the flexibility to deploy buffer server in any container still can be exercised once (or if) fault tolerance is supported by the buffer server.

Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: `bufferserver.purge()` is subject to NPE.

File: bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java [hunk 2]
```diff
@@ -178,19 +178,23 @@ public void reset()
 
   public void purge(final int baseSeconds, final int windowId)
```
Comment (vrozov) [inline bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java]: Remove `purge(int, int)` API (convert it to `purge(long)`).

File: engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java (hunk 1 above)
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: In StreamingContainerManager windowId to purge is decremented by 1. Please check if the same decrement needs to be applied here?

Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: In localcluster, buffer server is not started by default, making that change passes the test

File: engine/src/main/java/com/datatorrent/stram/StramLocalCluster.java [hunk 3]
```diff
@@ -72,7 +72,7 @@
   protected final StreamingContainerManager dnmgr;
   private final UmbilicalProtocolLocalImpl umbilical;
   private InetSocketAddress bufferServerAddress;
-  private boolean perContainerBufferServer;
+  private boolean perContainerBufferServer = true;
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/StramLocalCluster.java]: Why is it necessary to change the default value for perContainerBufferServer?

File: bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java [hunk 4]
```diff
@@ -176,21 +176,20 @@ public void reset()
     numberOfInMemBlockPermits.set(MAX_COUNT_OF_INMEM_BLOCKS - 1);
   }
 
-  public void purge(final int baseSeconds, final int windowId)
```
Comment (tweise) [inline bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java:179]: Are these changes necessary as part of this PR?

File: engine/src/main/java/com/datatorrent/stram/StramLocalCluster.java (hunk 3 above)
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/StramLocalCluster.java]: Same question as well, I think we should limit the changes to what is really needed.

File: engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java (hunk 1 above)
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: Please address the question regarding the windowId. If it isn't decremented, aren't we blowing away the checkpoint window?

File: engine/src/main/java/com/datatorrent/stram/StramLocalCluster.java (hunk 3 above)
Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/StramLocalCluster.java]: Agreed. I already changed it locally, will be in next update.

File: bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java (hunk 4 above)
Comment (sandeshh) [inline bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java:179]: That was the suggestion by @vrozov to remove the extra API. My preference is also to remove that, as it doesn't add much value.

File: engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java (hunk 1 above)
Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: Working on that, here are the current finding.
1. Previous purge - purged the windows before the committed window id. Do we need a data from the committed window?
2. Couldn't find the history for that in GitHub. That change is present atleast from 2013.

Next steps

Will conduct few test scenario to see why we need a window before the committed window in the buffer server.

Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: Wrote a sample application and tested this change by killing the containers. There was no issue and the recovery started from the expected window id ( committed + 1 ). Still not sure why we have to keep the window data of the committed window.

Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: Right, since that window was already materialized in the checkpointed state it shouldn't be needed. Have you checked that by looking where the subscriber starts consuming from?

Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: Subscriber request comes with committed window Id. Buffer Server sends the data from the next higher window.

Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: I would prefer to have a separate bug and PR for changing what window Id is safe to purge.

Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: I have created a jira for the extra window. Will update this PR accordingly.
https://issues.apache.org/jira/browse/APEXCORE-479

**Comment by sandeshh** [comment]
All the review comments has been addressed.

File: bufferserver/src/main/java/com/datatorrent/bufferserver/server/Server.java [hunk 5]
```diff
@@ -199,6 +199,13 @@ private void handlePurgeRequest(PurgeRequestTuple request, final AbstractLengthP
     }
   }
 
+  public void purge(long windowId)
+  {
+    for (DataList dataList: publisherBuffers.values()) {
```
Comment (vrozov) [inline bufferserver/src/main/java/com/datatorrent/bufferserver/server/Server.java:203]: The code is not thread safe as purge() will now be called from a different thread.

Comment (sandeshh) [inline bufferserver/src/main/java/com/datatorrent/bufferserver/server/Server.java:203]: Purge inside the _Block_ is not thread safe, but the Purge inside the _DataList_ calls the Block purge using Synchronized block. So it should be fine.

Comment (vrozov) [inline bufferserver/src/main/java/com/datatorrent/bufferserver/server/Server.java:203]: It is not thread safe to access HashMap (publisherBuffers) from 2 different threads in a case when one thread modifies HashMap (adds or removes entries in the HashMap). It is not fine as there is a race condition between purge() and handlePublisherRequest().

File: bufferserver/src/main/java/com/datatorrent/bufferserver/server/Server.java [hunk 6]
```diff
@@ -169,7 +168,7 @@ public String toString()
     return identity;
   }
 
-  private final HashMap<String, DataList> publisherBuffers = new HashMap<String, DataList>();
+  private final ConcurrentHashMap<String, DataList> publisherBuffers = new ConcurrentHashMap<String, DataList>();
```
Comment (vrozov) [inline bufferserver/src/main/java/com/datatorrent/bufferserver/server/Server.java]: Please use 1 for concurrency level instead of default 16. There is only one thread that modifies publisherBuffers. Use default 0.75f for the load factor and 1 as the initial capacity.

**Comment by vrozov** [comment]
@sandeshh Please rebase against latest master and I'll merge it.

File: bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java [hunk 7]
```diff
@@ -176,21 +176,20 @@ public void reset()
...
+  public void purge(long windowId)
```
Comment (gauravgopi123) [inline bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java]: should windowId be made final?

Comment (sandeshh) [inline bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java]: Will make that final and rebase the change.
